File: cpp/SwerveWithPathPlanner/src/main/cpp/Robot.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "Robot.hpp"
#include "LimelightHelpers.hpp"

#include "wpi/commands2/CommandScheduler.hpp"

Robot::Robot() {}

void Robot::RobotPeriodic() {
    timeAndJoystickReplay.Update();
    wpi::cmd::CommandScheduler::GetInstance().Run();

    /*
     * This example of adding Limelight is very simple and may not be sufficient for on-field use.
     * Users typically need to provide a standard deviation that scales with the distance to target
     * and changes with number of tags available.
     *
     * This example is sufficient to show that vision integration is possible, though exact implementation
     * of how to use vision should be tuned per-robot and to the team's specification.
     */
    if (USE_LIMELIGHT) {
        auto const driveState = container.drivetrain.GetState();
        auto const heading = driveState.Pose.Rotation().Degrees();
        auto const omega = driveState.Velocity.omega;

        LimelightHelpers::SetRobotOrientation("limelight", heading.value(), 0, 0, 0, 0, 0);
        auto llMeasurement = LimelightHelpers::getBotPoseEstimate_wpiBlue_MegaTag2("limelight");
        if (llMeasurement && llMeasurement->tagCount > 0 && wpi::units::math::abs(omega) < 2_tps) {
            container.drivetrain.AddVisionMeasurement(llMeasurement->pose, llMeasurement->timestampSeconds);
        }
    }
}

void Robot::DisabledInit() {}

void Robot::DisabledPeriodic() {}

void Robot::DisabledExit() {}

void Robot::AutonomousInit() {
    autonomousCommand = container.GetAutonomousCommand();

    if (autonomousCommand) {
        wpi::cmd::CommandScheduler::GetInstance().Schedule(autonomousCommand);
    }
}

void Robot::AutonomousPeriodic() {}

void Robot::AutonomousExit() {}

void Robot::TeleopInit() {
    if (autonomousCommand) {
        wpi::cmd::CommandScheduler::GetInstance().Cancel(autonomousCommand);
    }
}

void Robot::TeleopPeriodic() {}

void Robot::TeleopExit() {}

void Robot::UtilityInit() {
    wpi::cmd::CommandScheduler::GetInstance().CancelAll();
}

void Robot::UtilityPeriodic() {}

void Robot::UtilityExit() {}

#ifndef RUNNING_WPILIB_TESTS
int main() {
    return wpi::StartRobot<Robot>();
}
#endif

File: python/SwerveWithPathPlanner/robot.py
#!/usr/bin/env python3
#
# Copyright (c) FIRST and other WPILib contributors.
# Open Source Software; you can modify and/or share it under the terms of
# the WPILib BSD license file in the root directory of this project.
#

import commands2
from phoenix6 import HootAutoReplay, HootReplay

from robotcontainer import RobotContainer


class MyRobot(commands2.TimedCommandRobot):
    """
    Command v2 robots are encouraged to inherit from TimedCommandRobot, which
    has an implementation of robot_periodic which runs the scheduler for you
    """

    def __init__(self) -> None:
        """
        This function is run when the robot is first started up and should be used for any
        initialization code.
        """
        super().__init__()

        self._replay = HootReplay("./logs/example.hoot")

        self._autonomous_command: commands2.Command | None = None
        # Instantiate our RobotContainer.  This will perform all our button bindings, and put our
        # autonomous chooser on the dashboard.
        self._container = RobotContainer()

        # log and replay timestamp and Driver Station data
        self._time_and_ds_replay = (
            HootAutoReplay()
            .with_timestamp_replay()
            .with_driver_station_replay()
            .with_joystick_replay()
        )
        self.add_periodic(self._time_and_ds_replay.update, self.DEFAULT_PERIOD, -0.001)

    def robot_periodic(self) -> None:
        pass

    def disabled_init(self) -> None:
        pass

    def disabled_periodic(self) -> None:
        pass

    def disabled_exit(self) -> None:
        pass

    def autonomous_init(self) -> None:
        # This autonomous runs the autonomous command selected by your RobotContainer class.
        self._autonomous_command = self._container.get_autonomous_command()

        if self._autonomous_command:
            commands2.CommandScheduler.get_instance().schedule(self._autonomous_command)

    def autonomous_periodic(self) -> None:
        pass

    def autonomous_exit(self) -> None:
        pass

    def teleop_init(self) -> None:
        # This makes sure that the autonomous stops running when
        # teleop starts running. If you want the autonomous to
        # continue until interrupted by another command, remove
        # this line or comment it out.
        if self._autonomous_command:
            commands2.CommandScheduler.get_instance().cancel(self._autonomous_command)

    def teleop_periodic(self) -> None:
        pass

    def teleop_exit(self) -> None:
        pass

    def utility_init(self) -> None:
        # Cancels all running commands at the start of test mode
        commands2.CommandScheduler.get_instance().cancel_all()

    def utility_periodic(self) -> None:
        pass

    def utility_exit(self) -> None:
        pass
